[ONNX] Fix export for tuple outputs

KsenijaS opened the pull request:
Fix export for tuple outputs in tracing.

BowenBao committed:
[ONNX] Redesign onnx pass to enable shape type dependent pattern conversion - cont (#51795)

With the introduction of ONNX shape inference, shape and type are inferred on the fly as operators get converted from ATen to ONNX when running symbolic function. This resolves the shape/type requirement for the symbolic functions. The pre-onnx passes however, can not be supported by shape inference, since at that stage the operators in the graph are still ATen operators.

This PR is to update the design of ONNX pass, to enable a mechanism of capturing subgraphs of ATen operators of certain patterns, and convert them later, when shape/type information of upstream operators are available.

The new design will require pre-onnx passes that need shape/type to be written in two parts, encapsulation and conversion.

    The encapsulation part will find the nodes of patterns, like how pre-onnx passes were written previously. But instead of converting the nodes, it will encapsulate them into a sub-block of a new placeholder node. This part is called before onnx pass, so it runs before calling symbolic functions.

    The conversion part will be called inside the onnx pass. In onnx pass, run_symbolic_func will be called for each node in topological order. When it reaches the placeholder node, the conversion part will be invoked. It will convert the nodes inside the sub-block based on pattern. By that time, it will have shape/type of upstream operators available. After the conversion is complete, the placeholder node will be removed, and nodes inside its sub-block converted. Run_symbolic_func will be called for these nodes, and they will be converted from ATen operator to ONNX operator.

This PR includes several other fixes, listed below.
* ~~replace helper.cpp with onnx_utils.cpp for holding utility functions.~~
* fix EraseNumberTypes on Bool type, the code was outdated that back then Bool type doesn't exist.
* ~~enable onnx shape inference in export with parameter/initializer data.~~
* other code clean ups.
* fix insertion of identity nodes for loop opset 13 sequence output.

~~PR depends on #51603~~

jiafatom committed:
[ONNX] Remove the last Cast in pow symbolic_opset9 (#52646)

Fixes #52436
For opset 9 of onnx Pow, if X is int32, Y is float, we will cast back to int32 which is consistent with X type.
However, pytorch is still float. The aten graph sometimes does not bind with the type for operators,
we are fine with the float type and don't want to cast back.
Even if X, Y are int32, the resulting float32 and int32 makes no difference.

neginraoof committed:
[ONNX] Support inplace operations on inplace indexing (#52063)

* [ONNX] Fix for sequence of mutations in blocks (#51577)

Fixes consecutive mutations in a tensor inside blocks.
Also, support append and pop in blocks.

* Support inplace operations + indexing

* Clean up old pass for remove mutations

* Add loop test

* Fixes for set attr in loops

* Removing the new jit API flag

* [ONNX] Redesign onnx pass to enable shape type dependent pattern conversion - cont (#51795)

With the introduction of ONNX shape inference, shape and type are inferred on the fly as operators get converted from ATen to ONNX when running symbolic function. This resolves the shape/type requirement for the symbolic functions. The pre-onnx passes however, can not be supported by shape inference, since at that stage the operators in the graph are still ATen operators.

This PR is to update the design of ONNX pass, to enable a mechanism of capturing subgraphs of ATen operators of certain patterns, and convert them later, when shape/type information of upstream operators are available.

The new design will require pre-onnx passes that need shape/type to be written in two parts, encapsulation and conversion.

    The encapsulation part will find the nodes of patterns, like how pre-onnx passes were written previously. But instead of converting the nodes, it will encapsulate them into a sub-block of a new placeholder node. This part is called before onnx pass, so it runs before calling symbolic functions.

    The conversion part will be called inside the onnx pass. In onnx pass, run_symbolic_func will be called for each node in topological order. When it reaches the placeholder node, the conversion part will be invoked. It will convert the nodes inside the sub-block based on pattern. By that time, it will have shape/type of upstream operators available. After the conversion is complete, the placeholder node will be removed, and nodes inside its sub-block converted. Run_symbolic_func will be called for these nodes, and they will be converted from ATen operator to ONNX operator.

This PR includes several other fixes, listed below.
* ~~replace helper.cpp with onnx_utils.cpp for holding utility functions.~~
* fix EraseNumberTypes on Bool type, the code was outdated that back then Bool type doesn't exist.
* ~~enable onnx shape inference in export with parameter/initializer data.~~
* other code clean ups.
* fix insertion of identity nodes for loop opset 13 sequence output.

~~PR depends on #51603~~

* Fix after merge

* clang

* Fix clang

* Fix clang

* Fix warning message.

* Fixes for non-model param attributes

* Fix for caffe2

* Additional test

* clang

* Skip test for lower opsets

* fix clang-tidy

* Update init.cpp

* Update remove_inplace_ops_for_onnx.cpp

* Update remove_inplace_ops_for_onnx.cpp

* Update remove_inplace_ops_for_onnx.cpp

* Fix for clang formatting

Co-authored-by: Bowen Bao <bowbao@microsoft.com>

jiafatom committed:
[ONNX] Symbolic shape inference (#51481)

This PR did symbolic shape inference, in the onnx pass _jit_pass_onnx_graph_shape_type_inference.
It creates a singleton ConstantValueMap.
It leverages constant folding technique and did a per-op based handling for ConstantValueMap.
As a byproduct, it enables fold_if pass for dynamic axes cases, typically for faster-rcnn etc.

The core change is in `torch/csrc/jit/passes/onnx/shape_type_inference.cpp` and `torch/csrc/jit/passes/onnx/constant_map.cpp`.

We usually need copy tensor to store in the ConstantValueMap, otherwise the underlying value may change. I see this issue in (1) from_blob (2) get value from Constant node.

fatcat-z committed:
Update test cases generated by make_test() method to support running them in script mode. (#52748)

* Update tests for test_gru_* at this moment.

* Update flake8 error.

* Update tests for test_gru_* at this moment.

* Update flake8 error.

* Update test_gru_* test cases only.

* Fix flake8 issue.

* Fix flake8 issue on test.

* Still disable test cases created by make_test.

* Update code to fix issue 'AttributeError: 'RecursiveScriptModule' object has no attribute 'forward'' for test_elman_* test cases.

* Add script model support for test_lstm_* test cases.

hwangdeyu committed:
[ONNX] enable several script unit tests using new jit passes (#51722)

* enable scripting related unit test

* fix flaske

* enable more tests

* fix interpolate and BCElogits test

* fix interpolate test ci

* add test_interpolate_upsample opset comment

* add interpolate_upsample tracing support below opset 9

Co-authored-by: hwangdeyu <deyhuang@qq.com>

BowenBao committed:
Fix copy_ export (#53046)

Fixes export of torch.copy_

BowenBao committed:
[ONNX] Update assign output shape for nested structure and dict output (#52893)

Fixes dict output & nested tuple.

shubhambhokare1 committed:
[ONNX] Add repeat_interleave symbolic (#52855)

- Add support for aten::repeat_interleave
- NOTE: Also adds fix for cases with split op where input tensor sizes are not known but _outputs is provided

KsenijaS committed:
[ONNX] Update ONNX documentation (#51362)

Add information about .data field

KsenijaS committed:
trace tuple

BowenBao committed:
[ONNX] Support torch.isinf, torch.any and torch.all export to ONNX (#53328)

Supported for ONNX export after opset 10.
This is not exportable to opsets < 10 due to
1. onnx::IsInf is introduced in opset 10
2. onnx::Equal does not accept float tensor prior to opset 11

neginraoof committed:
[ONNX] fix export of embedding with padding_idx (#53053)

fix export of embedding with padding_idx

KsenijaS committed:
Merge branch 'onnx_ms_1' of https://github.com/pytorch/pytorch into tuple

BowenBao commented on torch/onnx/utils.py:
```diff
@@ -465,7 +465,10 @@ def _model_to_graph(model, args, verbose=False,
+            output_wrapped = []  # type: ignore
+            for el in torch_out:
+                if el is not None:
+                    output_wrapped.append(el)
```
- does it support None in tuple/list/nested?
- will this affect scripting?

neginraoof commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -1532,6 +1532,23 @@ def forward(self, x):
                       dynamic_axes={'input_1': [0, 1, 2],
                                     'output_1': [0, 1, 2]})
 
+    @disableScriptTest()
```
If you are disbling scripting, why are you testing with if/else? This is not reflected in the model in tracing.

KsenijaS committed:
update tracing tuple

shubhambhokare1 committed:
[ONNX] Improve index_put symbolic to handle singular Bool updates (#53690)

Adds support for cases where the updates to the index_put node is a single Bool value, such as the case shown below

```
mask[indices] = True
```

Fixes #53507

KsenijaS committed:
fix merge

KsenijaS committed:
fix merge conflict

KsenijaS committed:
fix merge conflict

KsenijaS committed:
fix merge conflict

BowenBao commented on torch/onnx/utils.py:
```diff
@@ -468,7 +481,7 @@ def _model_to_graph(model, args, verbose=False,
+            output_wrapped = _remove_none_list(torch_out)  # type: ignore
 
         output_tensors, out_desc = torch._C._jit_flatten(tuple(output_wrapped))
         torch._C._jit_pass_onnx_assign_output_shape(graph, output_tensors, out_desc, _onnx_shape_inference)
```
the `out_desc` afterwards is different, is that covered?

This doesn't feel compatible when support is added for None output from script module. Is there a design to be extensible for that?

shubhambhokare1 reviewed (commented):
Seems like PR #54623 is similar to this PR but handles ScriptModel cases as well

KsenijaS commented:
> Seems like PR #54623 is similar to this PR but handles ScriptModel cases as well
that's a good point. I can close my PR then. Although I am not sure if we should remove None from tuple outputs in scripting.
